### DISCUSSION TITLE
Change formatter to format code like in official documentation

### DISCUSSION THREAD

**Discussion by kyklish** [mark-wiemer/ahkpp#168] [Ideas]
What concern me:

1. formatter indent code in sharp commands `#If...`
2. formatter indent code inside labels (this is very useful)
3. formatter deletes indentation on empty strings

**1st** is bad: if user write only one `#If...` at top of the script, all below code will be indented. With this behavior formatter pushes user to use close `#If` command, but it not required by official documentation. I propose remove this indent.

**2nd** is very good, but breaks compatibly with other IDE (for example `AHK Studio`, it makes code like in official doc). I propose (and will do if we both agree) add setting `Disable indent inside labels` to extension's settings.

**3rd** is definitely must be configurable, for example `Keep indent of empty lines` setting.

Notifying @mark-wiemer

**Comment by mark-wiemer** [comment]
For now, it'd be best if our code copied the default formatting of SciTE4AutoHotkey, as I suspect that's the most popular formatter in current use. I'm no expert here, and official docs may be better, but it's hard to know exactly what formatting logic they use. As I continue work on #123, I hope to create a highly-configurable formatter that allows users to choose basically any setting they want.

Regarding **1st**, I think this should be a toggle. I don't want to change default formatting behavior (unless it's a bugfix) until version 3.0.0, which will close #123. I don't consider **1st** to be a bug, as this is a valuable behavior to know what code is under what if block (and easily collapse it)

**2nd** and **3rd** both make sense, and the fixes should be new settings like you've suggested :)

You're more than welcome to create three issues for this and open pull requests for the proposed solutions.

Thank you again for your contributions!

**Reply by kyklish** [reply]
Does SciTE4AutoHotkey has formatter for whole file? I didn't find it. It formats code during typing only?

**Reply by kyklish** [reply]
> this is a valuable behavior to know what code is under what if block

Official doc says: `#If` and `#IfWin...` are mutually exclusive; that is, only the most recent `#If` or `#IfWin...` will be in effect.
I think indent in this case irrelevant, but maybe can be optional.
I wrote test script and SciTE4AutoHotkey don't propose me indent after `#` directive.

**Reply by mark-wiemer** [reply]
Thanks for the update. I honestly haven't used SciTE4AutoHotkey much at all, it's just what seems to be popular. Do you know of other popular editors? If s4ahk isn't a good metric, we can make our own rules based on what community wants. The goal here is to be customizable though to account for other user preferences.

In that regard, let's stick with existing behavior by default, and add the option to remove the indent as a setting. This way, we won't disrupt users who are comfortable with this behavior, but we'll give the option to others to minimize the indentations. Does that sound reasonable?

**Reply by kyklish** [reply]
I agree with you :)

P.S.
I'm not push any changes or anything else, just discussing :), maybe some useful ideas come up from our chat.

**Comment by kyklish** [comment]
@mark-wiemer Is this a common style for `if-else`? Does a formatter have to support such a construction?

```ahk
if var1 {
    code
} if var2 {
    code
}
```

Relative to #261 PR. It will simplify code, if formatter will not support.

**Reply by mark-wiemer** [reply]
Did you mean `else if var2` in the middle? Usually 2 ifs would be separated:

```ahk
if var1 {
    code
}
if var2 {
    code
}
```

The closest I could see would be

```ahk
if var1 {
    code
} else if var2 {
    code
} else {
    code
}
```

**Reply by kyklish** [reply]
> Did you mean `} if var2 {` in the middle?

Yes.

> Usually 2 ifs would be separated.

Ok, I will drop support of such construction.

> 
> ```ahk
> } else if var2 {
> ```
This one is common style, I got implement it in formatter.
